### PR TITLE
Add cnt names

### PR DESCRIPTION
**Pull request by dhammika**
**Description:**

Adds shardIDs as connector container names. 

**Workflow steps:**
N/A

**Documentation links affected:**
N/A

**Notes for reviewers:**
N/A

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/flow/718)

### CODE REVIEW

File: go/bindings/build.go
Comment (psFried) [inline go/bindings/build.go:75]: I don't think `ContainerName` can be provided on `BuildArgs`. The problem is that a build may actually start _many_ containers, and they of course can't all have the same name. Builds will perform a `Validate` RPC for each capture and materialization that's involved in the build. There can be many of those, as a build can include arbitrary subsets of all the tasks in the control plane.

I think the solution here is to add the `ContainerName` parameter to the `CaptureDriverFn` and `MaterializeDriverFn`. Those functions each return a driver (backed by a running container) for a specific connector, and the task name, if applicable, should be known when those functions are called.

File: go/bindings/build.go
Comment (psFried) [inline go/bindings/build.go:146]: The name of the capture being validated here is `request.Capture`. I'm thinking this is where it could call a function that accepts the task and operation names, and returns a container name to pass to `CaptureDriverFn`

File: go/connector/run.go
Comment (psFried) [inline go/connector/run.go:153]: I'm realizing that there could be name collisions here. You could have two different Flow tasks, `foo/bar/baz` and `foo/bar_baz`, and both of those would result in the same `containerName`.

I'm not sure what we do about that. One option would be to base64 the whole provided name, but that might make it difficult to correlate the metrics later. Although, now that I look at the [cAdvisor docs](https://github.com/google/cadvisor/blob/master/docs/runtime_options.md#container-labels), I see that it's able to pass through labels! So maybe we can just base64 the provided name in order to get a valid container name, and add the original name as a label to allow the metrics to be correlated with the shard.

**Review by psFried — CHANGES_REQUESTED** [review]
Thanks, I think this is moving in the right direction, though there's more to do before it should merge. LMK what you think about the name sanitization.

I think in the meantime, I might try to temporarily hack in a UUID name for each container, just so I can change over to an explicit `docker stop` command. Then that name can be superceded by the one from this PR once it's ready.

File: go/connector/run.go
Comment (dhammika) [reply go/connector/run.go:153]: Nice, labeling is much more elegant! We can just keep docker assigned IDs and use labels for metrics tagging. 
This setup won't restrict duplicate containers from orchestrator misbehavior though. Tags will look something like, 
```
df6abb0ec43a    keen_ellis      shard=materialize/examples/stats/00000000-00000000,FLOW_RUNTIME_PROTOCOL=materialize,FLOW_TCP_PORT=2222,command=transactions
2b791ba97745    kind_shtern     FLOW_RUNTIME_PROTOCOL=materialize,FLOW_TCP_PORT=2222,command=transactions,shard=materialize/test/hello-world/ssh-postgresl/00000000-00000000
```

File: go/bindings/build.go
Comment (dhammika) [reply go/bindings/build.go]: Good point on `request.Capture`. I used that instead of `BuildArgs`. 
`ContainerName` is already passed to `CaptureDriverFn` and `MaterializeDriverFn`

File: go/bindings/build.go
Comment (dhammika) [reply go/bindings/build.go]: Patched below.

**Comment by dhammika** [comment]
I added SHA hashed names and CI is failing with duplicate name collisions. I'll check connector shutdown/eviction path.

**Review by psFried — APPROVED** [review]
I think this is looking pretty good. I'd like to address the couple of comments before merging, but I'm not actually certain that additional changes would be required, so LMK what you think.

File: go/flowctl-go/cmd-api-spec.go
Comment (dhammika) [reply go/flowctl-go/cmd-api-spec.go:39]: That flag is just a place holder in both `spec` and `discover` api calls. Gating those calls doesn't sound _right_ and honestly I'm not sure if different agents can make concurrent calls on the same image.

File: go/flowctl-go/cmd-api-discover.go
Comment (dhammika) [reply go/flowctl-go/cmd-api-discover.go:26]: Same as `spec` comment above.  I'd prefer patching caller (agent) later.

File: go/flowctl-go/cmd-api-spec.go
Comment (psFried) [reply go/flowctl-go/cmd-api-spec.go:39]: OK, sounds good to me. The runtime is where we have the most acute need, so I don't think spec or discover should block merging this PR

File: go/flowctl-go/cmd-api-discover.go
Comment (psFried) [reply go/flowctl-go/cmd-api-discover.go:26]: SGTM :+1:

**Comment by jgraettinger** [comment]
Not a review, but I did skim this and it looks 👍  to me.

**Pull request merged**
